### PR TITLE
Introduces FeatureModelTransaction API

### PR DESCRIPTION
**Pull request by vulder**

### CODE REVIEW

File: include/vara/Feature/FeatureModelTransaction.h [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+void addFeature(FeatureModel *FM, std::unique_ptr<Feature> NewFeature,
+                Feature *Parent = nullptr);
+
+//===----------------------------------------------------------------------===//
+//                    Transaction Implementation Details
+//===----------------------------------------------------------------------===//
+
+namespace detail {
+
+struct CopyTransactionMode {};
```
Comment (s9latimm) [inline include/vara/Feature/FeatureModelTransaction.h]:
```suggestion
class CopyTransactionMode {};
```
`CopyTransactionMode` defined as a struct here but previously declared as a class.

File: include/vara/Feature/FeatureModelTransaction.h [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+struct ModifyTransactionMode {};
```
Comment (s9latimm) [inline include/vara/Feature/FeatureModelTransaction.h]:
```suggestion
class ModifyTransactionMode {};
```
`ModifyTransactionMode` defined as a struct here but previously declared as a class.

File: lib/Feature/FeatureModel.cpp [hunk 3]
```diff
@@ -14,35 +14,32 @@ void FeatureModel::dump() const {
...
+}
+
+void FeatureModel::removeFeature(Feature &F) {
+  if (&F == Root) {
+    Root = nullptr;
   }
-  OrderedFeatures.insert(Value);
-  return true;
+  OrderedFeatures.remove(&F);
+  Features.erase(F.getName());
```
Comment (ChristianKaltenecker) [inline lib/Feature/FeatureModel.cpp:34]: Should we also ensure that the child feature no longer use this feature as parent? Or is this already the case?

File: include/vara/Feature/FeatureModelTransaction.h [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  FeatureModelTransaction(FeatureModel &FM) : TransactionBaseTy(FM) {}
+};
+
+using FeatureModelCopyTransaction =
+    FeatureModelTransaction<detail::CopyTransactionMode>;
+using FeatureModelModifyTransaction =
+    FeatureModelTransaction<detail::ModifyTransactionMode>;
+
+//===----------------------------------------------------------------------===//
+//                            Modifiction Helpers
```
Comment (ChristianKaltenecker) [inline include/vara/Feature/FeatureModelTransaction.h]:
```suggestion
//                            Modification Helpers
```

File: lib/Feature/FeatureModel.cpp (hunk 3 above)
Comment (vulder) [reply lib/Feature/FeatureModel.cpp:34]: Nope, the FM just offers private modifier methods that exactly change what requested, however, these are all private and only accessible to Modifications (after we refactored the builder). The Modification itself then guarantees and is tested to ensure that (e.g., AddFeature) does the correct thing, setting parent/child relation ships etc..

Furthermore, on top of that we have the transactions, which group to together a sequence of modifications and guarantee that these together produce a valid feature model.

**Review by s9latimm — APPROVED** [review]

**Review by ChristianKaltenecker — APPROVED** [review]

File: lib/Feature/FeatureModelParser.cpp [hunk 5]
```diff
@@ -98,6 +97,9 @@ bool FeatureModelXmlParser::parseConfigurationOption(xmlNode *Node,
       }
     }
   }
+  if (Name == "root") {
```
Comment (s9latimm) [inline lib/Feature/FeatureModelParser.cpp]:
```suggestion

  // XML has those names specified as root nodes
  if (Name == "root" || Name == "base") {
    FMB.setRootName(Name);
```

**Review by padupr — APPROVED** [review]

**Pull request merged**